# Assignment: Exercise 1.2
# Class: DSC 550
# Name: Wittlieff, Harlan
# Date: 12/5/2021

## 1. Load the dataset as a Pandas data frame.

In [49]:
# Import libraries
import pandas as pd
import numpy as np

# Load the dataset as Pandas data frame
df_videogames = pd.read_csv("Data\Video_Games_Sales_as_at_22_Dec_2016.csv")

## 2. Display the first ten rows of data.

In [50]:
df_videogames.head(10)

Name Platform  Year_of_Release         Genre  \
0                 Wii Sports      Wii           2006.0        Sports   
1          Super Mario Bros.      NES           1985.0      Platform   
2             Mario Kart Wii      Wii           2008.0        Racing   
3          Wii Sports Resort      Wii           2009.0        Sports   
4   Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                     Tetris       GB           1989.0        Puzzle   
6      New Super Mario Bros.       DS           2006.0      Platform   
7                   Wii Play      Wii           2006.0          Misc   
8  New Super Mario Bros. Wii      Wii           2009.0      Platform   
9                  Duck Hunt      NES           1984.0       Shooter   

  Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
0  Nintendo     41.36     28.96      3.77         8.45         82.53   
1  Nintendo     29.08      3.58      6.81         0.77         40.24   
2  Nintendo     15.68     12.76      3.79         3.29         35.52   
3  Nintendo     15.61     10.93      3.28         2.95         32.77   
4  Nintendo     11.27      8.89     10.22         1.00         31.37   
5  Nintendo     23.20      2.26      4.22         0.58         30.26   
6  Nintendo     11.28      9.14      6.50         2.88         29.80   
7  Nintendo     13.96      9.18      2.93         2.84         28.92   
8  Nintendo     14.44      6.94      4.70         2.24         28.32   
9  Nintendo     26.93      0.63      0.28         0.47         28.31   

   Critic_Score  Critic_Count User_Score  User_Count Developer Rating  
0          76.0          51.0          8       322.0  Nintendo      E  
1           NaN           NaN        NaN         NaN       NaN    NaN  
2          82.0          73.0        8.3       709.0  Nintendo      E  
3          80.0          73.0          8       192.0  Nintendo      E  
4           NaN           NaN        NaN         NaN       NaN    NaN  
5           NaN           NaN        NaN         NaN       NaN    NaN  
6          89.0          65.0        8.5       431.0  Nintendo      E  
7          58.0          41.0        6.6       129.0  Nintendo      E  
8          87.0          80.0        8.4       594.0  Nintendo      E  
9           NaN           NaN        NaN         NaN       NaN    NaN

## 3. Find the dimensions (number of rows and columns) in the data frame. What do these two numbers represent in the context of the data?

In [51]:
df_videogames.shape

(16719, 16)

This dataset contains 16 different metrics on 16,719 different video games.

## 4. Find the top five games by critic score.

In [52]:
df_videogames.sort_values(by = 'Critic_Score', axis = 0, ascending = False).head(5)

Name Platform  Year_of_Release     Genre  \
227   Tony Hawk's Pro Skater 2       PS           2000.0    Sports   
57         Grand Theft Auto IV      PS3           2008.0    Action   
51         Grand Theft Auto IV     X360           2008.0    Action   
5350               SoulCalibur       DC           1999.0  Fighting   
165         Grand Theft Auto V     XOne           2014.0    Action   

                 Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
227             Activision      3.05      1.41      0.02         0.20   
57    Take-Two Interactive      4.76      3.69      0.44         1.61   
51    Take-Two Interactive      6.76      3.07      0.14         1.03   
5350    Namco Bandai Games      0.00      0.00      0.34         0.00   
165   Take-Two Interactive      2.81      2.19      0.00         0.47   

      Global_Sales  Critic_Score  Critic_Count User_Score  User_Count  \
227           4.68          98.0          19.0        7.7       299.0   
57           10.50          98.0          64.0        7.5      2833.0   
51           11.01          98.0          86.0        7.9      2951.0   
5350          0.34          98.0          24.0        8.8       200.0   
165           5.48          97.0          14.0        7.9       764.0   

                    Developer Rating  
227   Neversoft Entertainment      T  
57             Rockstar North      M  
51             Rockstar North      M  
5350                    Namco      T  
165            Rockstar North      M

## 5. Find the number of video games in the data frame in each genre.

In [53]:
df_videogames.groupby('Genre')['Name'].count()

Genre
Action          3370
Adventure       1303
Fighting         849
Misc            1750
Platform         888
Puzzle           580
Racing          1249
Role-Playing    1500
Shooter         1323
Simulation       874
Sports          2348
Strategy         683
Name: Name, dtype: int64

## 6. Find the first five games in the data frame on the SNES platform.

In [54]:
df_videogames[df_videogames['Platform'] == 'SNES'].head(5)

Name Platform  Year_of_Release     Genre  \
18                      Super Mario World     SNES           1990.0  Platform   
56                  Super Mario All-Stars     SNES           1993.0  Platform   
71                    Donkey Kong Country     SNES           1994.0  Platform   
76                       Super Mario Kart     SNES           1992.0    Racing   
137  Street Fighter II: The World Warrior     SNES           1992.0  Fighting   

    Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
18   Nintendo     12.78      3.75      3.54         0.55         20.61   
56   Nintendo      5.99      2.15      2.12         0.29         10.55   
71   Nintendo      4.36      1.71      3.00         0.23          9.30   
76   Nintendo      3.54      1.24      3.81         0.18          8.76   
137    Capcom      2.47      0.83      2.87         0.12          6.30   

     Critic_Score  Critic_Count User_Score  User_Count Developer Rating  
18            NaN           NaN        NaN         NaN       NaN    NaN  
56            NaN           NaN        NaN         NaN       NaN    NaN  
71            NaN           NaN        NaN         NaN       NaN    NaN  
76            NaN           NaN        NaN         NaN       NaN    NaN  
137           NaN           NaN        NaN         NaN       NaN    NaN

## 7. Find the five publishers with the highest total global sales. Note: You will need to calculate the total global sales for each publisher to do this.

In [55]:
df_videogames[['Publisher', 'Global_Sales']].groupby('Publisher'). \
sum().sort_values(by = 'Global_Sales', axis = 0, ascending = False).head(5)

Global_Sales
Publisher                                
Nintendo                          1788.81
Electronic Arts                   1116.96
Activision                         731.16
Sony Computer Entertainment        606.48
Ubisoft                            471.61

## 8. Create a new column in the data frame that calculates the percentage of global sales from North America. Display the first five rows of the new data frame.

In [56]:
# Create a new column
df_videogames['Percent_North_America'] = df_videogames['NA_Sales'] / df_videogames['Global_Sales']

# Display the first five rows of the new data frame
df_videogames.head(5)

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count User_Score  User_Count Developer Rating  Percent_North_America  
0          51.0          8       322.0  Nintendo      E               0.501151  
1           NaN        NaN         NaN       NaN    NaN               0.722664  
2          73.0        8.3       709.0  Nintendo      E               0.441441  
3          73.0          8       192.0  Nintendo      E               0.476350  
4           NaN        NaN         NaN       NaN    NaN               0.359260

## 9. Find the number NaN entries (missing data values) in each column.

In [57]:
df_videogames.isnull().sum(axis = 0)

Name                        2
Platform                    0
Year_of_Release           269
Genre                       2
Publisher                  54
NA_Sales                    0
EU_Sales                    0
JP_Sales                    0
Other_Sales                 0
Global_Sales                0
Critic_Score             8582
Critic_Count             8582
User_Score               6704
User_Count               9129
Developer                6623
Rating                   6769
Percent_North_America       0
dtype: int64

## 10. Try to calculate the median user score of all the video games. You will likely run into an error because some of the user score entries are a non-numerical string that cannot be converted to a float. Find and replace this string with NaN and then calculate the median. Then, replace all NaN entries in the user score column with the median value.

In [58]:
# Find the different values in the user score column
df_videogames.User_Score.unique()

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'],
      dtype=object)

In [69]:
# text of 'tbd' identified, replace with NaN
df_videogames['User_Score'] = df_videogames['User_Score'].replace('tbd', np.nan)

# Convert User_Score to numeric
df_videogames['User_Score'] = pd.to_numeric(df_videogames['User_Score'])

In [75]:
# Calculate the mean
user_score_mean = df_videogames['User_Score'].mean()
print('Mean User_Score: ',user_score_mean)

Mean User_Score:  7.1250461133070315


In [76]:
# Replace all NaN entries in the user score column with the median value
df_videogames['User_Score'] = df_videogames['User_Score'].replace(np.nan, user_score_mean)

In [78]:
# The NaN values have been replaced
df_videogames.User_Score.unique()

array([8.        , 7.12504611, 8.3       , 8.5       , 6.6       ,
       8.4       , 8.6       , 7.7       , 6.3       , 7.4       ,
       8.2       , 9.        , 7.9       , 8.1       , 8.7       ,
       7.1       , 3.4       , 5.3       , 4.8       , 3.2       ,
       8.9       , 6.4       , 7.8       , 7.5       , 2.6       ,
       7.2       , 9.2       , 7.        , 7.3       , 4.3       ,
       7.6       , 5.7       , 5.        , 9.1       , 6.5       ,
       8.8       , 6.9       , 9.4       , 6.8       , 6.1       ,
       6.7       , 5.4       , 4.        , 4.9       , 4.5       ,
       9.3       , 6.2       , 4.2       , 6.        , 3.7       ,
       4.1       , 5.8       , 5.6       , 5.5       , 4.4       ,
       4.6       , 5.9       , 3.9       , 3.1       , 2.9       ,
       5.2       , 3.3       , 4.7       , 5.1       , 3.5       ,
       2.5       , 1.9       , 3.        , 2.7       , 2.2       ,
       2.        , 9.5       , 2.1       , 3.6       , 2.8    